models/TSD/poolings.py

```diff
@@ -0,0 +1,284 @@
+                group_size=1,
+                pooled_size=p.out_size,
+                part_size=0,
+                sample_per_part=4,
+                trans_std=0.1,
+                no_trans=False,
+                name='delta_c_pooled_feat'
+            ) # [1024, 256, 7, 7]
+
+            # roi_feat = X.roi_align(
+            #     feat_lvl,
+            #     rois=mx.sym.slice_axis(mx.sym.reshape(rois_lvl, (batch_size, image_rois, 5)), axis=2, begin=1, end=5),
+            #     out_size=p.out_size,
+            #     stride=stride,
+            #     name="delta_c_pooled_feat"
+            # )
+            # roi_feat = mx.sym.reshape(roi_feat, (batch_size*image_rois, -1, p.out_size, p.out_size))
+
+            fpn_roi_feats.append(roi_feat)
+        roi_feat = X.add_n(*fpn_roi_feats)
+
+        if p.fp16:
+            roi_feat = X.to_fp16(roi_feat, name="delta_c_roi_feat_to_fp16")
+
+        return roi_feat
+    
+    def get_roi_feature_test(self, conv_fpn_feat, rois, trans, image_rois, batch_size=1):
+        assert batch_size == 1
+        return self.get_roi_feature(conv_fpn_feat, rois, trans, image_rois, batch_size)
+
+
+class FPNRoIAlign_DeltaR(RoiExtractor):
+    def __init__(self, pRoi):
+        super().__init__(pRoi)
+
+    def get_roi_feature(self, conv_fpn_feat, rois, trans, image_rois, batch_size):
+        '''
+        conv_fpn_feat: list of feature maps
+        rois: [2, 512, 4]
+        trans: [1024, 2*7*7, 1, 1]
+        '''
+        p = self.p
+        rcnn_stride = p.stride
+        roi_canonical_scale = p.roi_canonical_scale
+        roi_canonical_level = p.roi_canonical_level
+
+        _trans = mx.sym.reshape(trans, [-1, image_rois, 2])
+        roi_group, offset_group = fpn_roi_assign_offset(mx.symbol, rois, _trans, rcnn_stride, roi_canonical_scale, roi_canonical_level)
+
+        rois_fpn = dict()
+        offset_fpn = dict()
+        for i, stride in enumerate(rcnn_stride):
+            rois_fpn["stride%s" % stride] = mx.sym.reshape(roi_group[i], [-1, 4])
+            offset_fpn["stride%s" % stride] = mx.sym.tile(
+                mx.sym.reshape(offset_group[i], [-1, 2, 1, 1]),
+                (1, 1, p.out_size, p.out_size)
+            )
+
+        if p.fp16:
+            for stride in rcnn_stride:
+                conv_fpn_feat["stride%s" % stride] = X.to_fp32(
+                    conv_fpn_feat["stride%s" % stride],
+                    name="fpn_stride%s_to_fp32"
+                )
+
+        batch_pad = mx.sym.reshape(mx.sym.repeat(mx.sym.arange(batch_size), repeats=image_rois, axis=0), [-1,1])
+
+        fpn_roi_feats = list()
+        for stride in rcnn_stride:
+            feat_lvl = conv_fpn_feat["stride%s" % stride]
+            rois_lvl = rois_fpn["stride%s" % stride]  # [1024, 4]
+            offset_lvl = offset_fpn["stride%s" % stride] # [1024, 2, 7, 7]
+            rois_lvl = mx.sym.concat(batch_pad, rois_lvl, dim=1)
+
+            roi_feat = mx.sym.contrib.DeformablePSROIPooling(
```

Comment:
Do not hard code parameters. For example, the output_dim should be the same as the number for channels of features in conv_fpn_feat.